File: 백준 16236 BFS 아기 상어.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class Program
{
    static int N;
    static int[,] board;
    static int[] dx = { -1, 1, 0, 0 };
    static int[] dy = { 0, 0, -1, 1 };
    static Shark shark = new Shark();

    class Position
    {
        public int X, Y, Dist;
        public Position(int x, int y, int dist)
        {
            X = x;
            Y = y;
            Dist = dist;
        }
    }

    class Shark
    {
        public int X, Y, Size = 2, Eaten = 0;
    }

    static void Main(string[] args)
    {
        N = int.Parse(Console.ReadLine());
        board = new int[N, N];

        for (int i = 0; i < N; i++)
        {
            var line = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();
            for (int j = 0; j < N; j++)
            {
                board[i, j] = line[j];
                if (board[i, j] == 9)
                {
                    shark.X = i;
                    shark.Y = j;
                    board[i, j] = 0;
                }
            }
        }

        int time = 0;
        while (true)
        {
            var fish = Bfs(shark);
            if (fish == null) break;

            board[fish.X, fish.Y] = 0;
            shark.X = fish.X;
            shark.Y = fish.Y;
            time += fish.Dist;
            shark.Eaten++;

            if (shark.Eaten == shark.Size)
            {
                shark.Size++;
                shark.Eaten = 0;
            }
        }

        Console.WriteLine(time);
    }

    static Position Bfs(Shark shark)
    {
        bool[,] visited = new bool[N, N];
        Queue<Position> queue = new Queue<Position>();
        List<Position> fishes = new List<Position>();

        queue.Enqueue(new Position(shark.X, shark.Y, 0));
        visited[shark.X, shark.Y] = true;

        while (queue.Count > 0)
        {
            var pos = queue.Dequeue();
            for (int i = 0; i < 4; i++)
            {
                int nx = pos.X + dx[i];
                int ny = pos.Y + dy[i];

                if (nx >= 0 && nx < N && ny >= 0 && ny < N && !visited[nx, ny] && board[nx, ny] <= shark.Size)
                {
                    visited[nx, ny] = true;
                    queue.Enqueue(new Position(nx, ny, pos.Dist + 1));

                    if (board[nx, ny] > 0 && board[nx, ny] < shark.Size)
                    {
                        fishes.Add(new Position(nx, ny, pos.Dist + 1));
                    }
                }
            }
        }

        if (fishes.Count == 0) return null;

        fishes.Sort((a, b) => a.Dist != b.Dist ? a.Dist.CompareTo(b.Dist) : a.X != b.X ? a.X.CompareTo(b.X) : a.Y.CompareTo(b.Y));
        return fishes[0];
    }
}

File: 백준 16236 BFS 아기 상어.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const N = parseInt(input[0], 10);
const board = input.slice(1, N + 1).map(line => line.split(' ').map(Number));
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];
let shark = { x: 0, y: 0, size: 2, eaten: 0 };

for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
        if (board[i][j] === 9) {
            shark.x = i;
            shark.y = j;
            board[i][j] = 0;
            break;
        }
    }
}

function bfs(shark) {
    const visited = Array.from(Array(N), () => Array(N).fill(false));
    const queue = [{ x: shark.x, y: shark.y, dist: 0 }];
    visited[shark.x][shark.y] = true;
    const fishes = [];

    while (queue.length > 0) {
        const { x, y, dist } = queue.shift();

        for (let i = 0; i < 4; i++) {
            const nx = x + dx[i];
            const ny = y + dy[i];

            if (nx >= 0 && nx < N && ny >= 0 && ny < N && !visited[nx][ny]) {
                if (board[nx][ny] <= shark.size) {
                    visited[nx][ny] = true;
                    queue.push({ x: nx, y: ny, dist: dist + 1 });

                    if (board[nx][ny] > 0 && board[nx][ny] < shark.size) {
                        fishes.push({ x: nx, y: ny, dist: dist + 1 });
                    }
                }
            }
        }
    }

    fishes.sort((a, b) => a.dist !== b.dist ? a.dist - b.dist : a.x !== b.x ? a.x - b.x : a.y - b.y);
    return fishes[0];
}

let time = 0;
while (true) {
    const fish = bfs(shark);
    if (!fish) break;

    board[fish.x][fish.y] = 0;
    shark.x = fish.x;
    shark.y = fish.y;
    time += fish.dist;
    shark.eaten++;

    if (shark.eaten === shark.size) {
        shark.size++;
        shark.eaten = 0;
    }
}

console.log(time);
